From 120905fbc48ac7658fac392113bf45e00880c456 Mon Sep 17 00:00:00 2001
From: Gabriel de Dietrich <gabriel.dietrich-de@nokia.com>
Date: Wed, 24 Feb 2010 16:55:15 +0100
Subject: [PATCH] Expand indicator would not be displayed after removal of a collapsed item's child

While setting the hasChildren property of QTreeViewItem, "collapsed" and
"not visible" were being mistaken.

Auto-test included.

Reviewed-by: Olivier
Task-number: QTBUG-7443
(cherry picked from commit 77670c3c0fdc3021356e212e94042a0b5a4f4f8c)
---
 src/gui/itemviews/qtreeview.cpp        |   13 +++++++++----
 tests/auto/qtreeview/tst_qtreeview.cpp |   14 ++++++++++++++
 2 files changed, 23 insertions(+), 4 deletions(-)

--- a/src/gui/itemviews/qtreeview.cpp
+++ b/src/gui/itemviews/qtreeview.cpp
@@ -3769,10 +3769,15 @@ void QTreeViewPrivate::rowsRemoved(const
         if (previousSibiling != -1 && after && model->rowCount(parent) == start)
             viewItems[previousSibiling].hasMoreSiblings = false;
 
-
-        updateChildCount(parentItem, -removedCount);
-        if (parentItem != -1 && viewItems.at(parentItem).total == 0)
-            viewItems[parentItem].hasChildren = false; //every children have been removed;
+        if (parentItem != -1) {
+            if (viewItems.at(parentItem).expanded) {
+                updateChildCount(parentItem, -removedCount);
+                if (viewItems.at(parentItem).total == 0)
+                    viewItems[parentItem].hasChildren = false; //every children have been removed;
+            } else if (viewItems[parentItem].hasChildren && !hasVisibleChildren(parent)) {
+                viewItems[parentItem].hasChildren = false;
+            }
+        }
         if (after) {
             q->updateGeometries();
             viewport->update();
